### ISSUE TITLE
[Python] Protobuf python generated pb2 files are without typing files

### ISSUE THREAD

**Issue by matejsp** [protocolbuffers/protobuf#16115]
Please generate with pyi files for python for builtin types.

Currently we have a lot of errors using those types with pylint and need to ignore checking:
```
'google.protobuf.timestamp_pb2' has no 'Timestamp' member .
```

<img width="394" alt="image" src="https://github.com/protocolbuffers/protobuf/assets/3091122/ce96b960-9cac-43f5-b5ab-05dbc80b8ed0">

**What version of protobuf and what language are you using?**
Version: 4.25.2 (from pypi)
Language: Python

**What operating system (Linux, Windows, ...) and version?**
MacOSX, Linux

**What runtime / compiler are you using (e.g., python version or gcc version)**
python 3.11.2

**What did you do?**
```
import "google/protobuf/timestamp.proto";

message SomeMessage {
  google.protobuf.Timestamp timestamp = 2;
}

python -m grpc_tools.protoc -I. --python_out=. --pyi_out=. --grpc_python_out=. *.proto
```

**What did you expect to see**
Protobuf should include pyi files and pass pylint validation when using timestampa type.

**What did you see instead?**
```
'google.protobuf.timestamp_pb2' has no 'Timestamp' member .
```

**Anything else we should know about your project / environment**
/

**Comment by Avasam** [comment]
This should be fairly easy to do using https://pypi.org/project/mypy-protobuf/
Here's how typeshed currently does it: https://github.com/python/typeshed/blob/main/scripts/generate_proto_stubs.sh

I'd love to see protobuf provide their own stubs (along with a [PEP 561 `py.typed` marker](https://peps.python.org/pep-0561/#packaging-type-information)). In the mean time you can install https://pypi.org/project/types-protobuf/

**Comment by matejsp** [comment]
@Avasam yes for mypy it works with mypy-protobuf, but not with pylint. We actually use both.

**Comment by Avasam** [comment]
Sounds like a pylint issue https://github.com/pylint-dev/pylint/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+pyi+in%3Atitle

**Comment by matejsp** [comment]
@Avasam pylint supports pyi just fine and we are using them in our own proto files. But when using timestamp from protobuf package it does not work for that specific type. This is the main reason for this ticket. To also generate typing pyi files. Because then it works out of the box with latest pylint.

For mypy yes it works with mypy-protobuf, but wouldn't it be better to just use python native typing option instead since it is already implemented and it works just bundled generated code is lacking them?

**Comment by matejsp** [comment]
This issue is still pending.

Python introduced pyi files with https://peps.python.org/pep-0561/ and it helps IDE, linters and other tools to correctly get type information from python files. Protobuf is still missing bundled pyi files for auto generated files (like timestamp) requiring third party plugins (like mypy-protobuf) instead of using recommended and official way.

**Comment by dyhn78** [comment]
@matejsp Isn't this what you are looking for?

> https://protobuf.dev/getting-started/pythontutorial/#compiling-protocol-buffers
> Protoc is also able to generate python stubs (.pyi) with --pyi_out.

**Comment by matejsp** [comment]
We do that already for our own generated python code from proto files.

But I cannot do that for google.protobuf.timestamp_pb2 since timestamp_pb2 is bundled from this repostiory python whl.

Basically this issue is about adding --pyi_out to the following code in this repo:
https://github.com/protocolbuffers/protobuf/blob/main/python/protobuf_distutils/protobuf_distutils/generate_py_protobufs.py

**Comment by dyhn78** [comment]
I see, so you want the builtin types to have pyi files as well 😅  understood.

**Comment by matejsp** [comment]
This issue is still relevant.

**Comment by jguamie** [comment]
@anandolee We need to make an associated Bazel change for this as well. Please speak to @haberman or @mkruskal-google about it.

**Comment by anandolee** [comment]
Looks like https://github.com/protocolbuffers/protobuf/commit/e3e56c72a9fafb64fd25853e183fffa2bd12b0b7 should fixed this issue. @matejsp can you check if this has been resolved?

**Comment by matejsp** [comment]
Based on the change it looks like it should work.

But when installing released protobuf last week, I don't see pyi files for _pb2 files :(
```
pip install protobuf==6.32.1
```

Currently we are using types_protobuf that contain: https://github.com/python/typeshed/tree/main/stubs/protobuf/google/protobuf 
but they are generated using mypy-protobuf and not by pyi_out.

**Comment by jguamie** [comment]
The PYI files aren't bundled with the Python release. @anandolee can you please fix this?

**Comment by mspiller** [comment]
pyi are not bundled in protobuf==7.34.1
